framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1029,6 +1029,42 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
         // check precondition annotations
         checkPreconditions(node, contractsUtils.getPreconditions(invokedMethodElement));
 
+        // Checks that the following rule is satisfied:
+        // The type on a constructor declaration must be a supertype of
+        // the return type of "super()" invocation within that constructor.
+        if (TreeUtils.isSuperCall(node)) {
+            TreePath path = atypeFactory.getPath(node);
+            MethodTree enclosingMethod = TreeUtils.enclosingMethod(path);
+            if (TreeUtils.isConstructor(enclosingMethod)) {
+                AnnotatedTypeMirror superType = atypeFactory.getAnnotatedType(node);
```

Comment:
I think we should extract this logic into a protected `checkSuperConstructorCall` helper method, similar to the other `checkXXX` methods that are called from here.
This will make type-system-dependent changes easier, makes this method shorter, and gives you more space to document the logic.